src/util/netif.cpp

```diff
@@ -0,0 +1,275 @@
+// Copyright (c) 2024 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or https://www.opensource.org/licenses/mit-license.php.
+
+#include <config/bitcoin-config.h> // IWYU pragma: keep
+
+#include <util/netif.h>
+
+#include <logging.h>
+#include <netbase.h>
+#include <util/check.h>
+#include <util/sock.h>
+#include <util/syserror.h>
+
+#if defined(__linux__) || defined(__FreeBSD__)
+
+#if defined(__linux__)
+#include <linux/rtnetlink.h>
+#elif defined(__FreeBSD__)
```

Comment:
Presumably this only exists in (the rather recent) FreeBSD >= 13.2? For older versions we need to use the same method as macOS? If so, then we probably need to check this stuff in configure.

Reply:
That said, it looks like FreeBSD 12 is pretty much unusable: https://forums.freebsd.org/threads/freebsd-12-2-stable-pkg-update-failed.92034/#post-640223

Reply:
If so i would prefer not supporting it at all for older FreeBSD. It's so rare compared to the other operating systems already, and most of the userbase will be setting their own firewall. But we indeed need to check that it doesn't cause a compilation error.

Reply:
Added a version check for at least 13.2.

Reply:
Right, that's good to know. However the current problem is that we can't compile this for *any* FreeBSD version. Looks like the  `MLMSG` etc macros don't work in C++, because `typeof` is used.

Reply:
That would probably work, but that'd be a really invasive workaround build system wise.

Reply:
Yes, please don't do that :)

Reply:
My expectation is that a `#define typeof __typeof__` around just the use of that macro on FreeBSD could fix it. Just haven't been able to try it out.

Reply:
This is the upstream fix: https://github.com/freebsd/freebsd-src/pull/1070.

The suggested workaround to define `typeof` is ok:

```diff
diff --git i/src/util/netif.cpp w/src/util/netif.cpp
index 845b8aed1d..1840974a83 100644
--- i/src/util/netif.cpp
+++ w/src/util/netif.cpp
@@ -9,18 +9,24 @@
 #include <logging.h>
 #include <netbase.h>
 #include <util/check.h>
 #include <util/sock.h>
 #include <util/syserror.h>
 
+#ifdef __FreeBSD__
+#include <osreldate.h>
+#endif
+
 // Linux and FreeBSD 13.2+
-#if defined(__linux__) || (defined(__FreeBSD__) && __FreeBSD_version >= 1302000)
+#if defined(__linux__) || __FreeBSD_version >= 1302000
 
 #if defined(__linux__)
 #include <linux/rtnetlink.h>
 #elif defined(__FreeBSD__)
+// Workaround https://github.com/freebsd/freebsd-src/pull/1070.
+#define typeof __typeof
 #include <netlink/netlink.h>
 #include <netlink/netlink_route.h>
 #endif
 
 static std::optional<CNetAddr> QueryDefaultGatewayImpl(sa_family_t family)
 {
@@ -77,14 +83,13 @@ static std::optional<CNetAddr> QueryDefaultGatewayImpl(sa_family_t family)
     } while (recv_result < 0 && (errno == EINTR || errno == EAGAIN));
     if (recv_result < 0) {
         LogPrintLevel(BCLog::NET, BCLog::Level::Error, "recv() from netlink socket: %s\n", NetworkErrorString(errno));
         return std::nullopt;
     }
 
-    size_t response_len = static_cast<size_t>(recv_result);
-    for (nlmsghdr* hdr = (nlmsghdr*)response; NLMSG_OK(hdr, response_len); hdr = NLMSG_NEXT(hdr, response_len)) {
+    for (nlmsghdr* hdr = (nlmsghdr*)response; NLMSG_OK(hdr, recv_result); hdr = NLMSG_NEXT(hdr, recv_result)) {
         rtmsg* r = (rtmsg*)NLMSG_DATA(hdr);
         int remaining_len = RTM_PAYLOAD(hdr);
 
         // Iterate over the attributes.
         rtattr *rta_gateway = nullptr;
         int scope_id = 0;
```

The above patch does two more things:
* `__FreeBSD_version` was not defined at the time it was checked, so `__FreeBSD_version >= 1302000` was always false. Include the header that defines it.
* Silence a warning about signed vs unsigned integer comparison.

I tested this by reverting the upstream fix - `s/__typeof/typeof` in my `/usr/include/netlink/netlink.h`. Confirming that it gives the typeof error, then applying to above workaround and confirming that it compiles.

Reply:
Thanks!

> Silence a warning about signed vs unsigned integer comparison.

Oh no. Pretty sure i had to do the `response_len` thing to work around another error about signed-unsigned integer comparison in the i686 CI run.

Reply:
On my Linux distro I have:

```cpp
struct nlmsghdr {          
        __u32           nlmsg_len;
...
#define NLMSG_OK(nlh,len) ((len) >= (int)sizeof(struct nlmsghdr) && \
                           (nlh)->nlmsg_len >= sizeof(struct nlmsghdr) && \
                           (nlh)->nlmsg_len <= (len))
```
So, the second argument of `NLMSG_OK()` - `len` is compared against 1. `(int)sizeof...` and 2. `...->nlmsg_len` (which is `__u32`). The only difference on FreeBSD is that in 2. it is compared against `(int)(_hdr)->nlmsg_len` - an explicitly cast `nlmsg_len` to `int` (`nlmsg_len` is `uint32_t` on FreeBSD).

I do not get a warning when compiling that code on Linux:
```cpp
    ssize_t response_len;
...
    for (nlmsghdr *hdr = (nlmsghdr*)response; NLMSG_OK(hdr, response_len); hdr = NLMSG_NEXT(hdr, response_len)) {
```
I am not sure why. I should get a warning about `__u32 <= ssize_t`, right?

Reply:
> I am not sure why. I should get a warning about __u32 <= ssize_t, right?

Comparing an unsigned type against a larger signed type does what one would expect: it gets casted to the larger signed type before comparison.
This is why on 64-bit platforms this is fine.
However on 32-bit platforms, `ssize_t` is 32 bit. So there it warns.

Reply:
I see, then defining `recv_result` as `int64_t` instead of `ssize_t` should be ok on all platforms?

Reply:
Will try